packages/cli/src/plugin.js

```diff
@@ -0,0 +1,228 @@
+  // we ignore any invalid names i.e. names without an '@'
+  const namespaces = Array.from(
+    new Set([
+      ...plugins
+        .filter((p) => p.package.startsWith('@redwoodjs/'))
+        .map((p) => p.package.split('/')[0])
+        .sort((a, b) =>
+          a.package > b.package ? 1 : b.package > a.package ? -1 : 0
+        ),
+      ...plugins
+        .filter((p) => !p.package.startsWith('@redwoodjs/'))
+        .map((p) => p.package.split('/')[0])
+        .sort((a, b) =>
+          a.package > b.package ? 1 : b.package > a.package ? -1 : 0
+        ),
+    ])
+  ).filter((ns) => ns.includes('@'))
+
+  // If the user is running a help command or no command was given
+  // we want to load all plugins for observability in the help output
+  const processArgv = process.argv.slice(2).join(' ')
+  const showingRootHelp =
+    processArgv === '--help' || processArgv === '-h' || processArgv === ''
+
+  // Filter the namespaces based on the command line args to
+  // reduce the number of plugins we need to load
+  const namespacesInUse = namespaces.filter(
+    (ns) => showingRootHelp || processArgv.includes(ns)
+  )
+  if (namespacesInUse.length === 0) {
+    // If no namespace is in use we're using the default @redwoodjs namespace which
+    // is just an empty string ''
+    namespacesInUse.unshift('@redwoodjs')
+  }
+
+  // TODO: We should have some mechanism to fetch the cache from an online or precomputed
+  // source this will allow us to have a cache hit on the first run of a command
+  let pluginCommandCache = {}
+  try {
+    pluginCommandCache = JSON.parse(
+      fs.readFileSync(
+        path.join(getPaths().generated.base, PLUGIN_CACHE_FILENAME)
+      )
+    )
+  } catch (_error) {
+    // No need to log this error, it's just a cache miss
+    // console.log(error)
+  }
+
+  // We filter plugins based on the first word which depends on if a namespace is in use
+  const firstWord = process.argv[2]?.includes('@')
+    ? process.argv[3]
+    : process.argv[2]
+  const showingNamespaceHelp =
+    firstWord === '--help' || firstWord === '-h' || firstWord === undefined
+
+  for (const namespace of namespacesInUse) {
+    // Get all the plugins for this namespace
+    const namespacedPlugins = new Set(
+      plugins.filter(
+        (p) => p.package.startsWith(namespace) && (p.enabled ?? true)
+      )
+    )
+    // Do nothing if there are no enabled plugins for this namespace
+    if (namespacedPlugins.size === 0) {
+      continue
+    }
+
+    // For help output we only show the root level commands which doesn't actually need
+    // to load any plugins other than the @redwoodjs namespaced ones
+    if (showingRootHelp && namespace !== '@redwoodjs') {
+      yargs.command({
+        command: `${namespace} <command>`,
+        describe: `${namespace} plugin commands`,
+        builder: () => {},
+        handler: () => {},
+      })
+      continue
+    }
```

Comment:
I'm having a hard time figuring out when this condition would be met. Check me on this, but showing root help only happens if a user does...

```
yarn rw --help
yarn rw -h
yarn rw
```

In the cases above, the namespace is `@redwoodjs`, right? So when would you show root help and the namespace not be `@redwoodjs`?

Reply:
I understand this now; it's to get help output for all the plugin commands